certora/specs/authorizations.spec

```diff
@@ -0,0 +1,263 @@
+
+    bool userHasRoleAfter = doesUserHaveRole(user, role);
+    assert (userHasRoleAfter != userHasRoleBefore =>
+            userChanged == user && roleChanged == role && userHasRoleAfter == enabledChanged);
+}
+
+// getRolesWithCapability mapping
+
+rule getRolesWithCapabilityChanging() {
+    env e;
+    method f; calldataarg args; bytes4 capability;
+    bytes32 rolesBefore = getRolesWithCapability(capability);
+
+    f(e, args);
+
+    bytes32 rolesAfter = getRolesWithCapability(capability);
+    assert (rolesAfter != rolesBefore =>
+            f.selector == setRoleCapability(uint8, bytes4, bool).selector);
+}
+
+rule setRoleCapabilityShouldChangeDoesRoleHaveCapability(uint8 role, bytes4 capability, bool enabled) {
+    env e;
+
+    setRoleCapability(e, role, capability, enabled);
+
+    bool roleHasCapabilityAfter = doesRoleHaveCapability(role, capability);
+    assert (roleHasCapabilityAfter == enabled);
+}
+
+rule doesRoleHaveCapabilityChangingMethod() {
+    env e;
+    method f; calldataarg args; uint8 role; bytes4 capability;
+    bool roleHasCapabilityBefore = doesRoleHaveCapability(role, capability);
+
+    f(e, args);
+
+    bool roleHasCapabilityAfter = doesRoleHaveCapability(role, capability);
+    assert (roleHasCapabilityAfter != roleHasCapabilityBefore =>
+            f.selector == setRoleCapability(uint8, bytes4, bool).selector);
+}
+
+rule doesRoleHaveCapabilityChangingArgs(uint8 roleChanged, bytes4 capabilityChanged, bool enabledChanged) {
+    env e;
+    uint8 role; bytes4 capability;
+    bool roleHasCapabilityBefore = doesRoleHaveCapability(role, capability);
+
+    setRoleCapability(e, roleChanged, capabilityChanged, enabledChanged);
+
+    bool roleHasCapabilityAfter = doesRoleHaveCapability(role, capability);
+    assert (roleHasCapabilityAfter != roleHasCapabilityBefore =>
+            roleChanged == role && capabilityChanged == capability && roleHasCapabilityAfter == enabledChanged);
+}
+
+
+// AUTHORIZATION FUNCTIONS
+
+// Compute the set of all the functions that need authorization.
+rule allFunctionsNeedingAuthorization() {
+    env e;
+    storage initialState = lastStorage;
+    method f; calldataarg args;
+
+    f@withrevert(e, args);
+    bool revertNormal = lastReverted;
+
+    env e_auth; bool enabled;
+    setPublicCapability(e_auth, utils.to_bytes4(f.selector), enabled) at initialState;
+    f@withrevert(e, args);
+    bool revertOwner = lastReverted;
+
+    assert ((revertOwner != revertNormal) =>
+            (f.selector == setOwner(address).selector) ||
+             f.selector == setPublicCapability(bytes4, bool).selector ||
+             f.selector == setUserRole(address, uint8, bool).selector ||
+             f.selector == setRoleCapability(uint8, bytes4, bool).selector ||
+             f.selector == transfer(address, uint256).selector ||
+             f.selector == transferFrom(address, address, uint256).selector ||
+             f.selector == mint(address, uint256).selector);
+}
```

Comment:
proving that only those functions need an authorization
but does not prove that each and every one of them has one

Reply:
True. Not sure how we would check that in a satisfying way though

Reply:
maybe by manually calling with someone that has been blocked/authorized and checking if it revert ?

Reply:
Alright let's do that, shouldn't be too long (but maybe a bit repetitive)

Reply:
Ok ping me when it's ready

Reply:
Consider yourself pinged ser